### ISSUE TITLE
Test failure printing no longer has test file line numbers, and is too verbose

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#18647]
I inserted one `@test 1 == 2` in a `@testset` loop in my package and this is the results:

``` jl
tree type: Test Failed
  Expression: 1 == 2
   Evaluated: 1 == 2
 in record(::Base.Test.DefaultTestSet, ::Base.Test.Fail) at ./test.jl:428
 in do_test(::Base.Test.Returned, ::Expr) at ./test.jl:281
 in macro expansion; at /home/kristoffer/.julia/v0.6/NearestNeighbors/test/test_inrange.jl:41 [inlined]
 in macro expansion; at ./test.jl:733 [inlined]
 in macro expansion; at /home/kristoffer/.julia/v0.6/NearestNeighbors/test/test_inrange.jl:0 [inlined]
 in macro expansion; at ./test.jl:733 [inlined]
 in macro expansion; at /home/kristoffer/.julia/v0.6/NearestNeighbors/test/test_inrange.jl:0 [inlined]
 in macro expansion; at ./test.jl:672 [inlined]
 in anonymous at ./<missing>:?
 in include_from_node1(::String) at ./loading.jl:532
 in include(::String) at ./sysimg.jl:14
 in include_from_node1(::String) at ./loading.jl:532
 in include(::String) at ./sysimg.jl:14
 in process_options(::Base.JLOptions) at ./client.jl:289
 in _start() at ./client.jl:353
Test Summary: | Pass  Fail  Total
  inrange     |   36     3     39
  metric      |   36     3     39
  tree type   |   12     1     13
  tree type   |   12     1     13
  tree type   |   12     1     13
ERROR: LoadError: LoadError: Some tests did not pass: 36 passed, 3 failed, 0 errored, 0 broken.
 in finish(::Base.Test.DefaultTestSet) at ./test.jl:495
 in macro expansion; at ./test.jl:679 [inlined]
 in anonymous at ./<missing>:?
 in include_from_node1(::String) at ./loading.jl:532
 in include(::String) at ./sysimg.jl:14
 in include_from_node1(::String) at ./loading.jl:532
 in include(::String) at ./sysimg.jl:14
 in process_options(::Base.JLOptions) at ./client.jl:289
 in _start() at ./client.jl:353
while loading /home/kristoffer/.julia/v0.6/NearestNeighbors/test/test_inrange.jl, in expression starting on line 2
while loading /home/kristoffer/.julia/v0.6/NearestNeighbors/test/runtests.jl, in expression starting on line 35
======================================[ ERROR: NearestNeighbors ]=======================================

failed process: Process(`/home/kristoffer/julia/usr/bin/julia -Cnative -J/home/kristoffer/julia/usr/lib/julia/sys.so --compile=yes --depwarn=yes --check-bounds=yes --code-coverage=none --color=yes --compilecache=yes /home/kristoffer/.julia/v0.6/NearestNeighbors/test/runtests.jl`, ProcessExited(1)) [1]

========================================================================================================
INFO: No packages to install, update or remove
ERROR: NearestNeighbors had test errors
 in #test#61(::Bool, ::Function, ::Array{AbstractString,1}) at ./pkg/entry.jl:749
 in (::Base.Pkg.Entry.#kw##test)(::Array{Any,1}, ::Base.Pkg.Entry.#test, ::Array{AbstractString,1}) at ./<missing>:0
 in (::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#test,Tuple{Array{AbstractString,1}}})() at ./pkg/dir.jl:31
 in cd(::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#test,Tuple{Array{AbstractString,1}}}, ::String) at ./file.jl:69
 in #cd#1(::Array{Any,1}, ::Function, ::Function, ::Array{AbstractString,1}, ::Vararg{Array{AbstractString,1},N}) at ./pkg/dir.jl:31
 in (::Base.Pkg.Dir.#kw##cd)(::Array{Any,1}, ::Base.Pkg.Dir.#cd, ::Function, ::Array{AbstractString,1}, ::Vararg{Array{AbstractString,1},N}) at ./<missing>:0
 in #test#3(::Bool, ::Function, ::String, ::Vararg{String,N}) at ./pkg/pkg.jl:258
 in test(::String, ::Vararg{String,N}) at ./pkg/pkg.jl:258
```

Is it possible that we could scrub away some of the stackframes here to make things more readable? The backtrace for the "Some tests did not pass" exception seems unnecessary and doesn't include anything related to the user. The backtrace for the actual test failure could maybe be changed so that all frames after the `test.jl` file gets removed?

**Comment by stevengj** [comment]
The backtraces in `make testall` failures from Julia Base are even worse.

**Comment by kshyatt** [comment]
This would be good to tackle after #17165 gets merged (which should happen soon!), I think.

**Comment by KristofferC** [comment]
This is a bit better now with https://github.com/JuliaLang/julia/pull/19751 and https://github.com/JuliaLang/julia/pull/19750. However, as @stevengj the `make` tests in Base are still pretty bad. I put two testsets that failed in one of the base test files and the resulting output is at https://gist.github.com/KristofferC/3e915887aa017470b72d7bed2f8956ff.

The expression for each failing test is printed three times. When it fails, at the end of the file, and at the end of the test run. We also print 5 backtraces. Two of them are where the test failed which are good ones. The other three, are noise from what I can see. What do people think, is it really necessary to print the expression for each failing test three times? From the posted gist I would estimate the noise/signal ratio to be about 2/3.

**Comment by StefanKarpinski** [comment]
Definitely printing less is much better. There should be one backtrace: the one we care about. The error should likewise be printed once.

**Comment by stevengj** [comment]
It's somewhat nice to have the errors reprinted at the very end of the test output, but maybe just the exception and the file/line where it was triggered?

**Comment by JeffBezanson** [comment]
The situation here is still pretty bad. Worst of all, no line number related to the failed test is shown. We used to print e.g. `while loading /tmp/julia/test/core.jl, in expression starting on line 99`. That line number is available within the `LoadError` thrown for an error during `include`.

Here's the output for a failed equality test:

```
Test (Worker)   | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
keywordargs: Test Failed
  Expression: 1 == 2
   Evaluated: 1 == 2
Stacktrace:
 [1] include_from_node1(::String) at ./loading.jl:539
 [2] include(::String) at ./sysimg.jl:14
 [3] macro expansion at /tmp/julia/test/testdefs.jl:18 [inlined]
 [4] macro expansion at ./test.jl:853 [inlined]
 [5] macro expansion at ./util.jl:378 [inlined]
 [6] macro expansion at /tmp/julia/test/testdefs.jl:0 [inlined]
 [7] anonymous at ./<missing>:?
Worker 1 failed running test keywordargs:
Some tests did not pass: 101 passed, 1 failed, 0 errored, 0 broken.keywordargs: Test Failed
  Expression: 1 == 2
   Evaluated: 1 == 2
Stacktrace:
 [1] record(::Base.Test.DefaultTestSet, ::Base.Test.Fail) at ./test.jl:564
 [2] (::##33#40)() at /tmp/julia/test/runtests.jl:170
 [3] cd(::##33#40, ::String) at ./file.jl:69
 [4] include_from_node1(::String) at ./loading.jl:539
 [5] include(::String) at ./sysimg.jl:14
 [6] process_options(::Base.JLOptions) at ./client.jl:305
 [7] _start() at ./client.jl:371

Test Summary: | Pass  Fail  Total
  Overall     |  101     1    102
    keywordargs |  101     1    102
    FAILURE
Error in testset keywordargs:
Test Failed
  Expression: 1 == 2
   Evaluated: 1 == 2
ERROR: LoadError: Test run finished with errors
while loading /tmp/julia/test/runtests.jl, in expression starting on line 39
```

The error is shown three times, which is silly. We should cut that down to 1 or 2. Next, I don't think it makes sense to show a stack trace for a failed equality test. It seems to just be a stack trace for the test system itself, which is not useful. Indeed, the system seems smart enough to omit the stack trace during the summary at the end.

Here's the output for an error during a test:

```
Test (Worker)   | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
keywordargs: Error During Test
  Test threw an exception of type MethodError
  Expression: f(1) == 1
  MethodError: no method matching +(::Int64, ::String)
  Closest candidates are:
    +(::Any, ::Any, !Matched::Any, !Matched::Any...) at operators.jl:424
    +(::T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8}, !Matched::T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8}) where T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8} at int.jl:32
    +(::Real, !Matched::Complex{Bool}) at complex.jl:235
    ...
  Stacktrace:
   [1] f(::Int64) at /tmp/julia/test/keywordargs.jl:2
   [2] include_from_node1(::String) at ./loading.jl:539
   [3] include(::String) at ./sysimg.jl:14
   [4] macro expansion at /tmp/julia/test/testdefs.jl:18 [inlined]
   [5] macro expansion at ./test.jl:853 [inlined]
   [6] macro expansion at ./util.jl:378 [inlined]
   [7] macro expansion at /tmp/julia/test/testdefs.jl:0 [inlined]
   [8] anonymous at ./<missing>:?
Worker 1 failed running test keywordargs:
Some tests did not pass: 101 passed, 0 failed, 1 errored, 0 broken.keywordargs: Error During Test
  Test threw an exception of type MethodError
  Expression: f(1) == 1
  MethodError: no method matching +(::Int64, ::String)
  Closest candidates are:
    +(::Any, ::Any, !Matched::Any, !Matched::Any...) at operators.jl:424
    +(::T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8}, !Matched::T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8}) where T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8} at int.jl:32
    +(::Real, !Matched::Complex{Bool}) at complex.jl:235
    ...
  Stacktrace:
   [1] f(::Int64) at /tmp/julia/test/keywordargs.jl:2
   [2] include_from_node1(::String) at ./loading.jl:539
   [3] include(::String) at ./sysimg.jl:14
   [4] macro expansion at /tmp/julia/test/testdefs.jl:18 [inlined]
   [5] macro expansion at ./test.jl:853 [inlined]
   [6] macro expansion at ./util.jl:378 [inlined]
   [7] macro expansion at /tmp/julia/test/testdefs.jl:0 [inlined]
   [8] anonymous at ./<missing>:?

Test Summary: | Pass  Error  Total
  Overall     |  101      1    102
    keywordargs |  101      1    102
    FAILURE
Error in testset keywordargs:
Error During Test
  Test threw an exception of type MethodError
  Expression: f(1) == 1
  MethodError: no method matching +(::Int64, ::String)
  Closest candidates are:
    +(::Any, ::Any, !Matched::Any, !Matched::Any...) at operators.jl:424
    +(::T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8}, !Matched::T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8}) where T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8} at int.jl:32
    +(::Real, !Matched::Complex{Bool}) at complex.jl:235
    ...
  Stacktrace:
   [1] f(::Int64) at /tmp/julia/test/keywordargs.jl:2
   [2] include_from_node1(::String) at ./loading.jl:539
   [3] include(::String) at ./sysimg.jl:14
   [4] macro expansion at /tmp/julia/test/testdefs.jl:18 [inlined]
   [5] macro expansion at ./test.jl:853 [inlined]
   [6] macro expansion at ./util.jl:378 [inlined]
   [7] macro expansion at /tmp/julia/test/testdefs.jl:0 [inlined]
   [8] anonymous at ./<missing>:?
ERROR: LoadError: Test run finished with errors
while loading /tmp/julia/test/runtests.jl, in expression starting on line 39
```

This is good since there is a stacktrace, but the full output is repeated 3 times.

Let's keep only the first copy of the error, and then maybe a shorter summary at the end.

**Comment by vtjnash** [comment]
We've changed and improved this several times. This issue is not helpful anymore.
